upstairs/src/lib.rs

```diff
@@ -4152,15 +4156,25 @@ impl Downstairs {
                     computed_hash
                 );
             }
+
             // no hash was correct
             Err(CrucibleError::HashMismatch)
         } else if !successful_decryption {
             // no hash + encryption context combination decrypted this block
             error!(log, "Decryption failed with correct hash");
             Err(CrucibleError::DecryptionError)
         } else {
-            // Ok!
-            Ok(valid_hash)
+            let Some(valid_hash) = valid_hash else {
```

Comment:
Nit: it's a little weird to have both `valid_hash: Option<u64>` and `successful_hash: bool`, since they reflect the same data (`successful_hash == valid_hash.is_some()`).

I'd suggest removing `successful_hash`, then restructure this conditional to something like

```rust
if let Some(valid_hash) = valid_hash {
    if successful_decryption {
        // .. happy case
    } else {
        // .. decryption failed
    }
} else {
    // ...missing hash case
}
```

Reply:
Agreed, done in c1c388b